File: Controlador/VentaControlador.cs
using System;
using System.Collections.Generic;
using System.Linq;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.Vista;
using Microsoft.Maui.Controls;

namespace punto_de_venta_C_.Controlador
{
    public class VentaControlador
    {
        private readonly ControladorPrincipal app;
        private VentaPanel panel;
        private List<ItemCarrito> carrito = new List<ItemCarrito>();
        private double descuento = 0;

        public VentaControlador(ControladorPrincipal app)
        {
            this.app = app;
        }

        public void setPanel(VentaPanel panel)
        {
            this.panel = panel;
        }

        public List<Producto> buscarProductos(string texto)
        {
            if (string.IsNullOrWhiteSpace(texto))
                return new List<Producto>();
            return app.getProductoBD().buscar(texto);
        }

        public async void agregarAlCarrito(Producto p)
        {
            foreach (ItemCarrito item in carrito)
            {
                if (item.getProducto().getId().Equals(p.getId()))
                {
                    double nuevaCant = item.getCantidad() + 1.0;
                    if (nuevaCant > p.getStock())
                    {
                        if (Application.Current?.MainPage != null)
                        {
                            await Application.Current.MainPage.DisplayAlert("Sin stock",
                                "Stock insuficiente. Disponible: " + (int)p.getStock(), "OK");
                        }
                        return;
                    }
                    item.incrementar(1.0);
                    notificarPanel();
                    return;
                }
            }
            if (p.getStock() <= 0)
            {
                if (Application.Current?.MainPage != null)
                {
                    await Application.Current.MainPage.DisplayAlert("Sin stock", "El producto no tiene existencia", "OK");
                }
                return;
            }
            carrito.Add(new ItemCarrito(p, 1.0));
            notificarPanel();
        }

        public void eliminarDelCarrito(string productoId)
        {
            carrito.RemoveAll(i => i.getProducto().getId().Equals(productoId));
            notificarPanel();
        }

        public async void setCantidad(string productoId, double nuevaCantidad)
        {
            foreach (ItemCarrito item in carrito)
            {
                if (item.getProducto().getId().Equals(productoId))
                {
                    nuevaCantidad = Math.Round(nuevaCantidad);
                    if (nuevaCantidad > item.getProducto().getStock())
                    {
                        if (Application.Current?.MainPage != null)
                        {
                            await Application.Current.MainPage.DisplayAlert("Sin stock",
                                "Stock insuficiente. Disponible: " + (int)item.getProducto().getStock(), "OK");
                        }
                        return;
                    }
                    item.setCantidad(nuevaCantidad);
                    break;
                }
            }
            notificarPanel();
        }

        public void setDescuento(double descuento)
        {
            this.descuento = Math.Max(0, descuento);
            notificarPanel();
        }

        public double getDescuento()
        {
            return descuento;
        }

        public double calcularSubtotal()
        {
            return carrito.Select(item => item.getSubtotal()).Sum();
        }

        public double calcularTotal()
        {
            double sub = calcularSubtotal();
            return Math.Round(Math.Max(0, sub - descuento) * 100.0) / 100.0;
        }

        public double calcularCambio(double recibido)
        {
            return Math.Max(0, recibido - calcularTotal());
        }

        public async void procesarCobro(string metodoPago, double montoRecibido)
        {
            if (carrito.Count == 0)
                return;
            double total = calcularTotal();
            if (montoRecibido < total && "Efectivo".Equals(metodoPago))
            {
                if (Application.Current?.MainPage != null)
                {
                    await Application.Current.MainPage.DisplayAlert("Importe insuficiente",
                        string.Format("El importe recibido (${0:0.00}) es menor al total (${1:0.00})", montoRecibido, total), "OK");
                }
                return;
            }
            Venta venta = new Venta(0, new List<ItemCarrito>(carrito), total, descuento, metodoPago, DateTime.Now,
                    app.getUsuarioActivo().getNombre());
            app.registrarVenta(venta);
            double cambio = calcularCambio(montoRecibido);
            
            TicketDialog ticket = new TicketDialog(venta, app.getConfig(), cambio, descuento, montoRecibido);
            await ticket.mostrar();
            
            carrito.Clear();
            descuento = 0;
            if (panel != null)
                panel.limpiar();
        }

        private void notificarPanel()
        {
            if (panel != null)
                panel.refrescarCarrito(carrito, calcularTotal(), descuento);
        }

        public List<ItemCarrito> getCarrito()
        {
            return carrito;
        }
    }
}

File: Controlador/VentaControlador.py
from Modelo.ItemCarrito import ItemCarrito
from Modelo.Venta import Venta
from datetime import datetime
from PyQt6.QtWidgets import QMessageBox

class VentaControlador:
    def __init__(self, app):
        self.app = app
        self.panel = None
        self.carrito = []
        self.descuento = 0.0

    def setPanel(self, panel):
        self.panel = panel

    def buscarProductos(self, texto):
        if not texto or not str(texto).strip():
            return []
        return self.app.getProductoBD().buscar(texto)

    def agregarAlCarrito(self, p):
        for item in self.carrito:
            if item.getProducto().getId() == p.getId():
                nuevaCant = item.getCantidad() + 1.0
                if nuevaCant > p.getStock():
                    QMessageBox.warning(None, "Sin stock", f"Stock insuficiente. Disponible: {int(p.getStock())}")
                    return
                item.setCantidad(nuevaCant)
                self.notificarPanel()
                return
        
        if p.getStock() <= 0:
            QMessageBox.warning(None, "Sin stock", "El producto no tiene existencia")
            return
            
        item = ItemCarrito()
        item.setProducto(p)
        item.setCantidad(1.0)
        self.carrito.append(item)
        self.notificarPanel()

    def eliminarDelCarrito(self, productoId):
        self.carrito = [i for i in self.carrito if i.getProducto().getId() != productoId]
        self.notificarPanel()

    def setCantidad(self, productoId, nuevaCantidad):
        for item in self.carrito:
            if item.getProducto().getId() == productoId:
                nuevaCantidad = round(nuevaCantidad)
                if nuevaCantidad > item.getProducto().getStock():
                    QMessageBox.warning(None, "Sin stock", f"Stock insuficiente. Disponible: {int(item.getProducto().getStock())}")
                    return
                item.setCantidad(nuevaCantidad)
                break
        self.notificarPanel()

    def setDescuento(self, descuento):
        self.descuento = max(0.0, descuento)
        self.notificarPanel()

    def getDescuento(self):
        return self.descuento

    def calcularSubtotal(self):
        return sum(item.getProducto().getPrecio() * item.getCantidad() for item in self.carrito)

    def calcularTotal(self):
        sub = self.calcularSubtotal()
        return round(max(0.0, sub - self.descuento), 2)

    def calcularCambio(self, recibido):
        return max(0.0, recibido - self.calcularTotal())

    def procesarCobro(self, metodoPago, montoRecibido):
        if not self.carrito:
            return
        total = self.calcularTotal()
        if montoRecibido < total and str(metodoPago).lower() == "efectivo":
            QMessageBox.warning(None, "Importe insuficiente", f"El importe recibido (${montoRecibido:.2f}) es menor al total (${total:.2f})")
            return
            
        venta = Venta()
        venta.setId(0)
        venta.setItems(list(self.carrito))
        venta.setTotal(total)
        venta.setDescuento(self.descuento)
        venta.setMetodoPago(metodoPago)
        venta.setFecha(datetime.now())
        venta.setCajero(self.app.getUsuarioActivo().getNombre() if self.app.getUsuarioActivo() else "SISTEMA")
        
        self.app.registrarVenta(venta)
        cambio = self.calcularCambio(montoRecibido)
        
        from Vista.TicketDialog import TicketDialog
        parent_frame = self.app.getVentanaPrincipal().getFrame() if self.app.getVentanaPrincipal() and hasattr(self.app.getVentanaPrincipal(), 'getFrame') else None
        ticket = TicketDialog(parent_frame, venta, self.app.getConfig(), cambio, self.descuento, montoRecibido)
        ticket.mostrar()
        
        self.carrito.clear()
        self.descuento = 0.0
        if self.panel is not None:
            self.panel.limpiar()

    def notificarPanel(self):
        if self.panel is not None:
            self.panel.refrescarCarrito(self.carrito, self.calcularTotal(), self.descuento)

    def getCarrito(self):
        return self.carrito
